zrml/currencies/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+        ) -> result::Result<Self::Balance, DispatchError> {
+            T::Currencies::repatriate_reserved(currency_id, slashed, beneficiary, value, status)
+        }
+    }
+
+    impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
+        type CurrencyId = CurrencyIdOf<T>;
+        type Balance = BalanceOf<T>;
+        fn minimum_balance(currency_id: Self::CurrencyId) -> Self::Balance {
+            T::Currencies::minimum_balance(currency_id)
+        }
+        fn total_issuance(currency_id: Self::CurrencyId) -> Self::Balance {
+            T::Currencies::total_issuance(currency_id)
+        }
+        fn total_balance(currency_id: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
+            T::Currencies::total_balance(currency_id, who)
+        }
+        fn free_balance(currency_id: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
+            T::Currencies::free_balance(currency_id, who)
+        }
+        fn ensure_can_withdraw(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            T::Currencies::ensure_can_withdraw(currency_id, who, amount)
+        }
+        fn transfer(
+            currency_id: Self::CurrencyId,
+            from: &T::AccountId,
+            to: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            Accounts::<T>::insert(currency_id, to, ());
+            T::Currencies::transfer(currency_id, from, to, amount)
+        }
+        fn deposit(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            Accounts::<T>::insert(currency_id, who, ());
+            T::Currencies::deposit(currency_id, who, amount)
+        }
+        fn withdraw(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            T::Currencies::withdraw(currency_id, who, amount)
+        }
+
+        fn can_slash(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> bool {
+            T::Currencies::can_slash(currency_id, who, amount)
+        }
+        fn slash(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> Self::Balance {
+            T::Currencies::slash(currency_id, who, amount)
+        }
+    }
+
+    impl<T: Config> ZeitgeistAssetManager<T::AccountId> for Pallet<T> {
+        fn accounts_by_currency_id(
+            currency_id: Self::CurrencyId,
+        ) -> Result<(usize, Vec<T::AccountId>), DispatchError> {
+            let accounts: Vec<T::AccountId> =
+                Accounts::<T>::iter_prefix(currency_id).map(|(k2, _v)| k2).collect();
+            Ok((accounts.len(), accounts))
+        }
```

Comment:
**Overall:** I mean: what is the point of using another storage double map, when we iterate through the whole original map [here ](https://github.com/zeitgeistpm/zeitgeist/blob/89e36c13995135edbf6ac6b2cc9e025051c3ce44/primitives/src/traits/zeitgeist_multi_reservable_currency.rs#L53-L63) anyways? Or is the plan to remove the whole iterations of https://github.com/zeitgeistpm/zeitgeist/blob/89e36c13995135edbf6ac6b2cc9e025051c3ce44/primitives/src/traits/zeitgeist_multi_reservable_currency.rs#L45-L64 ? And then we only use `zrml_currencies` in the future.

Reply:
yes only remove_all will be used from orml_token, accounts_by.. will be used from zrml_currencies

Reply:
Avoid using `.collect()`, because this requires an expensive heap allocation. Rather return the Iterator itself.
https://nnethercote.github.io/perf-book/iterators.html

Reply:
it seems I should be using drain_prefix() as it provides iterator and removes the key values as well so , I could avoid allocation required while .collect()